lasagne/layers/special.py

```diff
@@ -15,6 +16,57 @@
 ]
 
 
+class ExpressionLayer(Layer):
+    """
+    This layer provides boilerplate for a custom layer that applies a
+    simple transformation to the input.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+
+    function : callable
+        A function to be applied to the output of the previous layer.
+
+    output_shape : None, callable, tuple, or 'auto'
+        Specifies the output shape of this layer. If a tuple, this fixes the
+        output shape for any input shape (the tuple can contain None if some
+        dimensions may vary). If a callable, it should return the calculated
+        output shape given the input shape. If None, the output shape is
+        assumed to be the same as the input shape. If 'auto', an attempt will
+        be made to automatically infer the correct output shape.
+    """
+    def __init__(self, incoming, function, output_shape=None, **kwargs):
+        super(ExpressionLayer, self).__init__(incoming, **kwargs)
+
+        if output_shape is not None:
+            if output_shape == 'auto':
+                self._output_shape = 'auto'
+            elif hasattr(output_shape, '__call__'):
+                self.get_output_shape_for = output_shape
+            else:
+                self._output_shape = tuple(output_shape)
+
+        self.function = function
+
+    def get_output_shape_for(self, input_shape):
+        try:
+            output_shape = self._output_shape
+            if output_shape == 'auto':
+                input_shape = (0 if s is None else s for s in input_shape)
+                X = theano.tensor.alloc(0, *input_shape)
+#                output_shape = tuple(self.function(X).shape.eval())
+                output_shape = self.function(X).shape.eval()
+                output_shape = tuple(s if s else None for s in output_shape)
+            return output_shape
+        except AttributeError:
+            return input_shape
```

Comment:
I'm not sure about placing the `except AttributeError` so far from the line that's meant to cause it. I'd prefer either:
```python
try:
    output_shape = self._output_shape
except AttributeError:
    return input_shape
else:
    ...
```
or setting `self._output_shape` to `None` in the constructor if `output_shape` is `None`. That way, `_output_shape` is always defined when `get_output_shape_for` wasn't overwritten and can simply be checked for.